stablehlo/transforms/VhloLegalizeToStablehlo.cpp

```diff
@@ -464,6 +464,198 @@ SpecialResult convertSpecial(const OpConversionPattern<VhloOpTy>& pattern,
+  return attr && attr.getValue() == value;
+}
+
+bool isEmptyArray(Attribute vhloAttr) {
+  auto attr = vhloAttr.dyn_cast_or_null<vhlo::ArrayV1Attr>();
+  return attr && attr.getValue().empty();
+}
+
+bool isEmptyString(Attribute vhloAttr) {
+  auto attr = vhloAttr.dyn_cast_or_null<vhlo::StringV1Attr>();
+  return attr && attr.getValue().empty();
+}
+
+bool isEmptyTensor(Attribute vhloAttr) {
+  auto attr = vhloAttr.dyn_cast_or_null<DenseElementsAttr>();
+  return attr && attr.getNumElements() == 0;
+}
+
+bool isEnum(Attribute vhloAttr, Attribute value) { return vhloAttr == value; }
+
+bool isInteger(Attribute vhloAttr, int64_t value) {
+  auto attr = vhloAttr.template dyn_cast_or_null<vhlo::IntegerV1Attr>();
+  return attr && attr.getValue().getSExtValue() == value;
+}
+
+bool isSplatArray(Attribute vhloAttr, Attribute splatValue) {
+  auto attr = vhloAttr.dyn_cast_or_null<vhlo::ArrayV1Attr>();
+  return attr && llvm::all_of(attr.getValue(), [&](Attribute attr) {
+           return attr == splatValue;
+         });
+}
+
+template <typename T>
+bool isSplatTensor(const ConversionPattern& pattern, Attribute vhloAttr,
+                   T splatValue) {
+  auto attr = convertGeneric(vhloAttr, pattern.getTypeConverter())
+                  .template dyn_cast_or_null<DenseElementsAttr>();
+  return attr && attr.isSplat() &&
+         attr.template getSplatValue<T>() == splatValue;
+}
+
+bool isString(Attribute vhloAttr, StringRef value) {
+  auto attr = vhloAttr.dyn_cast_or_null<vhlo::StringV1Attr>();
+  return attr && attr.getValue() == value;
+}
+
+// TODO: Also validate attributes before removing them.
+// Current logic assumes that these attributes are valid, but that might not
+// necessarily be the case because VHLO doesn't have verifiers.
+template <typename VhloOpTy>
+LogicalResult removeDefaults(const OpConversionPattern<VhloOpTy>& pattern,
+                             VhloOpTy vhloOp,
+                             SmallVector<NamedAttribute>& vhloAttrs) {
+  if constexpr (std::is_same<VhloOpTy, vhlo::FuncOpV1>::value) {
+    if (isString(vhloOp.getSymVisibilityAttr(), "public"))
+      eraseAttrs(vhloAttrs, "sym_visibility");
+    if (isEmptyArray(vhloOp.getArgAttrsAttr()))
+      eraseAttrs(vhloAttrs, "arg_attrs");
+    if (isEmptyArray(vhloOp.getResAttrsAttr()))
+      eraseAttrs(vhloAttrs, "res_attrs");
+  }
+  if constexpr (std::is_same<VhloOpTy, vhlo::AllGatherOpV1>::value ||
+                std::is_same<VhloOpTy, vhlo::AllReduceOpV1>::value ||
+                std::is_same<VhloOpTy, vhlo::AllToAllOpV1>::value ||
+                std::is_same<VhloOpTy, vhlo::CollectivePermuteOpV1>::value ||
+                std::is_same<VhloOpTy, vhlo::ReduceScatterOpV1>::value) {
+    if (isInteger(vhloOp.getChannelIdAttr(), 0))
+      eraseAttrs(vhloAttrs, "channel_id");
+    // AllToAllOpV1 and CollectivePermuteOpV1 don't have use_global_device_ids,
+    // so we cannot use getUseGlobalDeviceIds here.
```

Comment:
Should we split these two into a different constexpr?

Reply:
Done - this will add a little bit of additional type safety.